File: src/commonMain/kotlin/io/smallibs/lang/nethra/ast/Term.kt
package io.smallibs.lang.nethra.ast

/*
    n in Ident
    i in Int
    c in Char

    e ::=
        Type_i                -- Type at level i
        n                     -- Variable
        C(i:e)                -- Constructor                            ???

        i                     -- Integer literal
        c                     -- Character literal

        Π(n:e).e   Π{n:e}.e   -- Dependant function type
        λ(n).e     λ{n}.e     -- Function
        e e        e {e}      -- Application

        Σ(n:e).e              -- Dependant pair type
        e , e                 -- Pair
        fst e                 -- Left projection
        snd e                 -- Right Projection

        e + e                 -- Disjunction
        inl e                 -- Left injection
        inr e                 -- Right injection
        case e e e            -- Catamorphism

        rec(n).e              -- Recursion
        fold (rec(n).e) e     -- Fold recursive type
        unfold (rec(n).e) e   -- Unfold recursive type

        e ∈ e                 -- Term inhabits Type

        ?n                    -- Hole for inference
*/

sealed class Term<C>(open val context: C? = null) {

    sealed interface Literal {
        data class IntLit(val value: Int) : Literal
        data class CharLit(val value: Char) : Literal
    }

    data class Type<C>(val level: Int = 0) : Term<C>()
    data class Data<C>(val value: String, val type: Term<C>) : Term<C>()
    data class Id<C>(val value: String) : Term<C>()

    data class Lit<C>(val literal: Literal) : Term<C>()

    data class Pi<C>(val n: String, val bound: Term<C>, val body: Term<C>, val implicit: Boolean = false) : Term<C>()
    data class Lambda<C>(val n: String, val body: Term<C>, val implicit: Boolean = false) : Term<C>()
    data class Apply<C>(val abstraction: Term<C>, val argument: Term<C>, val implicit: Boolean = false) : Term<C>()

    data class Sigma<C>(val n: String, val bound: Term<C>, val body: Term<C>) : Term<C>()
    data class Couple<C>(val lhd: Term<C>, val rhd: Term<C>) : Term<C>()
    data class Fst<C>(val term: Term<C>) : Term<C>()
    data class Snd<C>(val term: Term<C>) : Term<C>()

    data class Disjunction<C>(val lhd: Term<C>, val rhd: Term<C>) : Term<C>()
    data class Inl<C>(val term: Term<C>) : Term<C>()
    data class Inr<C>(val term: Term<C>) : Term<C>()
    data class Case<C>(val term: Term<C>, val left: Term<C>, val right: Term<C>) : Term<C>()

    data class Rec<C>(val self: String, val body: Term<C>) : Term<C>()
    data class Fold<C>(val type: Rec<C>, val term: Term<C>) : Term<C>()
    data class Unfold<C>(val type: Rec<C>, val term: Term<C>) : Term<C>()

    data class Inhabit<C>(val term: Term<C>, val type: Term<C>) : Term<C>()

    data class Hole<C>(val value: String, var term: Term<C>? = null) : Term<C>()
}

File: ocaml/lib/nethra/lang/render/term.ml
open Format

let kind ppf (level, _) =
  if level == 0 then fprintf ppf "type" else fprintf ppf "type%d" level

let int ppf (value, _) = fprintf ppf "%d" value
let char ppf (value, _) = fprintf ppf "'%c'" value
let string ppf (value, _) = fprintf ppf "\"%s\"" value

let id ppf (value, initial, _) =
  match initial with
  | Some value -> fprintf ppf "%s" value
  | None -> fprintf ppf "%s" value

let pi ppf render (n, bound, body, implicit, _) =
  fprintf ppf
    (if implicit then "Π{%s:%a}.%a" else "Π(%s:%a).%a")
    n render bound render body

let lambda ppf render (n, body, implicit, _) =
  fprintf ppf (if implicit then "λ{%s}.(%a)" else "λ(%s).(%a)") n render body

let apply ppf render (abstraction, argument, implicit, _) =
  fprintf ppf
    (if implicit then "%a {%a}" else "%a %a")
    render abstraction render argument

let sigma ppf render (n, bound, body, _) =
  fprintf ppf "Σ(%s:%a).%a" n render bound render body

let pair ppf render (left, right, _) =
  fprintf ppf "(%a,%a)" render left render right

let fst ppf render (term, _) = fprintf ppf "fst %a" render term
let snd ppf render (term, _) = fprintf ppf "snd %a" render term

let sum ppf render (left, right, _) =
  fprintf ppf "%a | %a" render left render right

let inl ppf render (term, _) = fprintf ppf "inl %a" render term
let inr ppf render (term, _) = fprintf ppf "inr %a" render term

let case ppf render (term, left, right, _) =
  fprintf ppf "case %a %a %a" render term render left render right

let mu ppf render (n, body, _) = fprintf ppf "μ(%s).(%a)" n render body
let fold ppf render (term, _) = fprintf ppf "fold %a" render term
let unfold ppf render (term, _) = fprintf ppf "unfold %a" render term

let hole ppf render (value, reference, _) =
  match !reference with
  | Some value -> fprintf ppf "%a" render value
  | None -> fprintf ppf "%s" value

let rec render ppf t =
  Nethra_ast.Term.Destruct.fold ~kind:(kind ppf) ~int:(int ppf) ~char:(char ppf)
    ~string:(string ppf) ~id:(id ppf) ~pi:(pi ppf render)
    ~lambda:(lambda ppf render) ~apply:(apply ppf render)
    ~sigma:(sigma ppf render) ~pair:(pair ppf render) ~fst:(fst ppf render)
    ~snd:(snd ppf render) ~sum:(sum ppf render) ~inl:(inl ppf render)
    ~inr:(inr ppf render) ~case:(case ppf render) ~mu:(mu ppf render)
    ~fold:(fold ppf render) ~unfold:(unfold ppf render) ~hole:(hole ppf render)
    t
